### ISSUE TITLE
Moving away from CMocka

### ISSUE THREAD

**Issue by brockho** [numbbo/coco#1625]
CMocka causes more and more problems in our setting. We should therefore move away from this library. Question: which other library shall we use? We want: simple and popular.

see also https://stackoverflow.com/questions/65820/unit-testing-c-code#65845

**Comment by nikohansen** [comment]
Unfortunately this hasn't been solved yet. Now the circleci C tests fail due to
```
COPY	code-experiments/test/unit-test/lib/linux_ubuntu/libcmocka.so -> code-experiments/test/unit-test/libcmocka.so
COPY	code-experiments/build/c/coco.c -> code-experiments/test/unit-test/coco.c
EXPAND	code-experiments/src/coco.h to code-experiments/test/unit-test/coco.h
MAKE	clean in code-experiments/test/unit-test
CHDIR	/root/coco/code-experiments/test/unit-test
COPY	Makefile.in -> Makefile
rm -f unit_test.o unit_test

MAKE	all in code-experiments/test/unit-test
CHDIR	/root/coco/code-experiments/test/unit-test
COPY	Makefile.in -> Makefile
gcc -c -g -ggdb -std=c89 -pedantic -Wall -Wextra -Wstrict-prototypes -Wshadow -Wno-sign-compare -Wconversion  -o unit_test.o unit_test.c 
gcc -g -ggdb -std=c89 -pedantic -Wall -Wextra -Wstrict-prototypes -Wshadow -Wno-sign-compare -Wconversion  -o unit_test unit_test.o -L. -lm -lcmocka  
/usr/bin/x86_64-linux-gnu-ld: skipping incompatible ./libcmocka.so when searching for -lcmocka

RUN	./unit_test in code-experiments/test/unit-test
ERROR: return value=-11
[==========] Running 2 test(s).
[ RUN      ] test_logger_biobj_evaluate
COCO INFO: Results will be output to folder exdata/default

COCO INFO: 28.04.18 10:40:09, d=2, running: f23.done
[       OK ] test_logger_biobj_evaluate
[ RUN      ] UH��H��PH�}��
Could not run the test - check test fixtures
[  ERROR   ] UH��H��PH�}��
[==========] 2 test(s) run.
[  PASSED  ] 1 test(s).
[==========] Running 2 test(s).
[ RUN      ] test_coco_archive

COCO INFO: 28.04.18 10:40:09, d=2, running: f23.[       OK ] test_coco_archive
[ RUN      ] UH��H��@H�}ȿ
Could not run the test - check test fixtures
[  ERROR   ] UH��H��@H�}ȿ
[==========] 2 test(s) run.
[  PASSED  ] 1 test(s).
[==========] Running 6 test(s).
[ RUN      ] test_coco_is_nan
[       OK ] test_coco_is_nan
[ RUN      ] UH��H��H�}�H��
Could not run the test - check test fixtures
[  ERROR   ] UH��H��H�}�H��

Exited with code 255
```

**Comment by nikohansen** [comment]
@dtusar @islandblue @ttusar , do you see a reasonable pathway to get the C tests in a more stable shape? IMHO, our C code risks to become entirely unmaintainable in not so far future if we don't do anything.

**Comment by ttusar** [comment]
We could try to use [MinUnit](https://github.com/siu/minunit/blob/master/README.md). It seems to be really simple (everything is given in a single header, so no need to compile any libraries), but should be powerful enough for our needs. That is, if the unit tests in C have not gotten considerably more complex lately or need some special functionalities (I don't remember any tests needing mocking "back in my day"). Also, people seem to [like it](https://stackoverflow.com/a/65962).

**Comment by ttusar** [comment]
Just a reminder in case we do choose to use MinUnit: it is licensed under the MIT License (we need to acknowledge this in our licence).

**Comment by ttusar** [comment]
MinUnit does not follow the C89 standard, because it uses the `snprintf` function and the `__file__` macro. I can get around the missing `snprintf` function quite easily by using [its portable implementation](https://www.ijs.si/software/snprintf/), but it's trickier for the `__file__` macro. The easyest fix is to remove the `-pedantic` compiler option for the unit tests. Then everything works, but the tests (of course) do not follow the C89 standard strictly. 

Now the question is, can we live with that? The core code would still be compiled with the `-pedantic`  option, as would be the case with the integration tests, this would only affect the unit tests. 

Other than that, `MinUnit` is easy to use and it should not take a lot of effort to shift from `cmocka` to `MinUnit`.

**Comment by ttusar** [comment]
Forgot to mention the not-so-easy (and rather ungy) fix: `MinUnit` uses the `__func__` macro only for outputting where the test failed. So we could technically remove this information and it should work...

**Comment by ttusar** [comment]
Well, we still have integration tests that are compiled and run under the C89 standard, so my guess would be no.

**Comment by ttusar** [comment]
When I tested this solution on the slaves other problems came up, so I ended up removing some functionalities from `MinUnit` that caused its incompliance with the C89 standard (see below for more info). 

```c
/**
 * A C89-compliant instance of MinUnit (a minimal unit testing framework for C).
 *
 * The following changes were made to ensure the implementation is compliant with the C89 standard
 * (some functionalities have been removed):
 *   - No timers
 *   - No usage of __func__
 *   - snprintf replaced with the write_last_message function based on coco_strdupf and coco_error
 */
````

Now everything works. This issue can be closed once #1779 is accepted.
